sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/dynamodb/DynamoDBIO.java

```diff
@@ -504,17 +517,20 @@ public void processElement(ProcessContext context) throws Exception {
       }
 
       private Map<String, AttributeValue> extractDeduplicateKeyValues(WriteRequest request) {
+        List<String> deduplicationKeys = spec.getDeduplicateKeys();
+        Map<String, AttributeValue> attributes = Collections.emptyMap();
+
         if (request.putRequest() != null) {
-          return request.putRequest().item().entrySet().stream()
-              .filter(entry -> spec.getDeduplicateKeys().contains(entry.getKey()))
-              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+          attributes = request.putRequest().item();
         } else if (request.deleteRequest() != null) {
-          return request.deleteRequest().key().entrySet().stream()
-              .filter(entry -> spec.getDeduplicateKeys().contains(entry.getKey()))
-              .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        } else {
-          return Collections.emptyMap();
+          attributes = request.deleteRequest().key();
+        }
+
+        if (attributes.isEmpty() || deduplicationKeys.isEmpty()) {
+          return attributes;
         }
+
+        return ImmutableMap.copyOf(filterKeys(attributes, deduplicationKeys::contains));
```

Comment:
Changed the logic to using `Maps.filterKeys` so the intend is rather obvious compared to before:
```java
attributes.entrySet().stream()
  .filter(entry -> spec.getDeduplicateKeys().contains(entry.getKey()))
  .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
```